File: extractor.js
const fs = require('fs');

function extract(filename, startStr, endStr) {
    const html = fs.readFileSync(filename, 'utf-8');
    const lines = html.split(/\r?\n/);
    
    let startIdx = -1;
    let endIdx = -1;
    
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].includes(startStr) && startIdx === -1) {
            startIdx = i;
        } else if (lines[i].includes(endStr) && startIdx !== -1 && endIdx === -1) {
            endIdx = i;
        }
    }
    
    if (startIdx !== -1 && endIdx !== -1) {
        return lines.slice(startIdx, endIdx).join('\n') + '\n';
    }
    return null;
}

const oldCalc = extract('index_13dbc13.html', 'function calculateModel(', 'function getGrade(');
const oldApp = extract('index_13dbc13.html', 'function renderApp(', 'async function runBacktest()');
const oldBook = extract('index_13dbc13.html', 'function changeBook(', 'function formatDateStr(');

const newCalc = extract('index.html', 'function calculateModel(', 'function getGrade(');
const newApp = extract('index.html', 'function renderApp(', 'async function runBacktest()');
const newBook = extract('index.html', 'function changeBook(', 'function formatDateStr(');

console.log('Old App lines:', oldApp ? oldApp.split('\n').length : 0);
console.log('New App lines:', newApp ? newApp.split('\n').length : 0);

let html = fs.readFileSync('index.html', 'utf-8');

// Replace using literal string replacement
html = html.replace(newCalc, oldCalc);
html = html.replace(newApp, oldApp);
html = html.replace(newBook, oldBook);

fs.writeFileSync('index.html', html, 'utf-8');

console.log("Replacement successful.");

File: extractor.py
import sys

def extract(filename, start_str, end_str):
    with open(filename, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    start_idx = -1
    end_idx = -1
    
    for i, line in enumerate(lines):
        if start_str in line and start_idx == -1:
            start_idx = i
        elif end_str in line and start_idx != -1 and end_idx == -1:
            end_idx = i
            
    if start_idx != -1 and end_idx != -1:
        return "".join(lines[start_idx:end_idx])
    return None

old_calc = extract('index_13dbc13.html', 'function calculateModel(', 'function getGrade(')
old_app = extract('index_13dbc13.html', 'function renderApp(', 'async function runBacktest()')
old_book = extract('index_13dbc13.html', 'function changeBook(', 'function formatDateStr(')

new_calc = extract('index.html', 'function calculateModel(', 'function getGrade(')
new_app = extract('index.html', 'function renderApp(', 'async function runBacktest()')
new_book = extract('index.html', 'function changeBook(', 'function formatDateStr(')

with open('index.html', 'r', encoding='utf-8') as f:
    html = f.read()

html = html.replace(new_calc, old_calc)
html = html.replace(new_app, old_app)
html = html.replace(new_book, old_book)

with open('index.html', 'w', encoding='utf-8') as f:
    f.write(html)

print("Replacement successful.")
